Core/MMKV_IO.cpp

```diff
@@ -643,6 +647,28 @@ bool MMKV::setDataForKey(MMBuffer &&data, MMKVKey_t key, bool isDataHolder) {
     {
         auto itr = m_dic->find(key);
         if (itr != m_dic->end()) {
+
+            // JUST demo
+            auto basePtr = (uint8_t *) (m_file->getMemory()) + Fixed32Size;
+            MMBuffer oldBuffer = itr->second.toValueDataMMBuffer(basePtr);
+            if (oldBuffer.length() > 0) {
+                // TODO: compare in place
+                uint8_t *buffer = new uint8_t[oldBuffer.length()];
+                CodedOutputData output(buffer, oldBuffer.length());
+                try {
+                    writeValueData(data, isDataHolder, &output);
```

Comment:
This new approach might work. But you might consider a cleaner way of doing it. By saying cleaner, I mean, by this point, it's valueless to continue the DataHolder logic/flag anymore, for the efficiency gained with DataHolder is almost lost in these calls `new[]` / `writeValueData()` and `delete[]`.

Reply:
For example, you might consider switching to continue with `isDataHolder`=`false` (by encoding the value into a solid MMBuffer, with size encoded) if an existing value of the same key is found.

Reply:
Very appreciated for detailed comment and advice. Some comments will be fixed later, because I want to use it for incoming code review.

Frankly, now I want to discuss with you, two interesting methods: one is encoding, the other one is decoding

FOR encoding, we encoding new data to new whole MMBuffer, compare it with old MMBuffer.
FOR decoding, we decode old MMBuffer to a new raw  MMBuffer(if needed, according to `isDataHolder` )， then compare with the new raw MMBuffer.

codes is near by `#ifdef COMPARE_METHOD1`

`Encoding`, sometimes, need memory allocation.
`Decoding`, does not need memory allocation, so it have better performance.

In addition, I think this two method are equivalent for somehow, So I prefer use `decoding` method for good performance

I am just worried about one thing: encoding means more frequent memory usage.

Thanks a lot

Reply:
If we can compare values without encoding it, no doubt it's better. But I'm afraid method 2 is wrong as well, or it creates more problems than it trying to solve (such as the introduction of -1 length MMBuffer). I think method 2 is a good direction, you should keep on that. With careful consideration & robust coding, it would work well.
You could look into `MMKV::getValueSize()` for some hint about how to robust-check existing value, especially when it doesn't contain a valid MMBuffer inside.

Reply:
`MMKV::getValueSize()` surprised me, I read the source code, feel that it looks not being so rigorous.

I just am very curious: if there exists a dataset which is NOT a holder value, BUT be recognized as a holder value

MAYBE  we can find or calculate one by `bool MMKV::set(const vector<string> &v, MMKVKey_t key, uint32_t expireDuration)`

Reply:
Vector<string> is in fact stored as MMBuffer. What is your point?

Reply:
I misunderstand.

**new changes:**
1、change `readData` from "<=" to "==" for compareBeforeSet, imitating logic of `MMKV::getValueSize()`
2、MMBuffer(-1) is replaced by `exception`, if it doesn't contain a valid MMBuffer inside, no  need to compare anymore
3、exception tests are added in demo.cpp
4、`getDataForKey` in method[2] is replaced by `it->second` directly, to reduce an extra finding from `std::map`

codes for method[1] will be removed when when think method[2] are good enough.

thanks a lot